# BASIC PYTHON 6

## PANDAS

> <span style="color:blue">Pandas</span> adalah library Python yang fokus pada data analysis. Sering disebut dengan library data analysis karena banyak modul untuk mengolah data.

> Dokumentasi: https://pandas.pydata.org/about/


### Series dan Data Frame

> Pandas memiliki 2 tipe data untuk versi terbaru, yaitu Series dan Data Frame.

> Dokumentasi: https://pandas.pydata.org/docs/getting_started/index.html


### INSTALL & IMPORT Pandas

In [1]:
# install pandas (anaconda sudah menyediakan pandas)
# !pip install pandas

In [2]:
# import library pandas
import pandas as pd
import numpy as np

# untuk menjadi data analyst harus bisa ini juga
# import matplotlib.pyplot as plt
# import seaborn as sns

In [3]:
# cek dokumentasi
pd?

Type:        module
String form: <module 'pandas' from 'C:\\ProgramData\\Anaconda3\\lib\\site-packages\\pandas\\__init__.py'>
File:        c:\programdata\anaconda3\lib\site-packages\pandas\__init__.py
Docstring:  
pandas - a powerful data analysis and manipulation library for Python

**pandas** is a Python package providing fast, flexible, and expressive data
structures designed to make working with "relational" or "labeled" data both
easy and intuitive. It aims to be the fundamental high-level building block for
doing practical, **real world** data analysis in Python. Additionally, it has
the broader goal of becoming **the most powerful and flexible open source data
analysis / manipulation tool available in any language**. It is already well on
its way toward this goal.

Main Features
-------------
Here are just a few of the things that pandas does well:

  - Easy handling of missing data in floating point as well as non-floating
    point data.
  - Size mutability: columns can be ins

### 1. Membaca Data Tabular

![](https://pandas.pydata.org/docs/_images/02_io_readwrite.svg)

In [4]:
# membaca data csv dari link
df = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv")

In [5]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [6]:
# membaca data csv dari lokal
pd.read_csv("data/titanic.csv")

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [7]:
# cek tipe data
type(df)

pandas.core.frame.DataFrame

### 2. Membuat Dataframe Object

> Dokumentasi: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html#pandas.DataFrame

In [8]:
# contoh membuat dataframe
# cara 1, buat dataframe kosong dan list secara terpisah,
# lalu tempelkan list ke dataframe

contoh1 = pd.DataFrame()

In [9]:
# buat list
list1 = ["Bambang", "Irfan", "Firman", "Andik"]
list2 = ["Pamungkas", "Bachdim", "Utina", "Vermansyah"]
print(list1)
print(list2)

['Bambang', 'Irfan', 'Firman', 'Andik']
['Pamungkas', 'Bachdim', 'Utina', 'Vermansyah']


In [10]:
# masukkan list ke dataframe
contoh1["nama_depan"] = list1
contoh1["nama_belakang"] = list2

In [11]:
contoh1

nama_depan nama_belakang
0    Bambang     Pamungkas
1      Irfan       Bachdim
2     Firman         Utina
3      Andik    Vermansyah

In [12]:
# contoh 2
contoh2 = pd.DataFrame(list(zip(list1, list2)),
                       columns=["nama_depan", "nama_belakang"])

In [13]:
contoh2

nama_depan nama_belakang
0    Bambang     Pamungkas
1      Irfan       Bachdim
2     Firman         Utina
3      Andik    Vermansyah

In [14]:
# cara 3, buat dataframe dengan dictionary
dict1 = {"nama_depan": list1,
         "nama_belakang": list2}

print(dict1)

{'nama_depan': ['Bambang', 'Irfan', 'Firman', 'Andik'], 'nama_belakang': ['Pamungkas', 'Bachdim', 'Utina', 'Vermansyah']}


In [15]:
# buat dataframe
contoh3 = pd.DataFrame(dict1)

In [16]:
# lihat perubahan dataframe
contoh3

nama_depan nama_belakang
0    Bambang     Pamungkas
1      Irfan       Bachdim
2     Firman         Utina
3      Andik    Vermansyah

In [17]:
# untuk melihat seriesnya, kita dapat melakukan sliciing pada nama kolomnya
print(contoh3["nama_belakang"])

0     Pamungkas
1       Bachdim
2         Utina
3    Vermansyah
Name: nama_belakang, dtype: object


In [18]:
print(type(contoh3["nama_belakang"]))

<class 'pandas.core.series.Series'>


### 3. Attributes, Indexing, dan Slicing

> Dalam penggunaanya, kita dapat menggunakan beberapa fungsi attribute pada pandas untuk mengetahui lebih lanjut mengenai informasi dari data yang kita miliki, diantaranya adalah `head()`, `tail()`, `index`, `column`, `dtypes`, `info()`, `values`, `ndim`, dan `size`.

In [19]:
# cara mengakses data paling awal, default 5
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [20]:
# mengakses data paling akhir, default 5
df.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

In [21]:
# melihat index
df.index

RangeIndex(start=0, stop=891, step=1)

In [22]:
# melihat kolom
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [23]:
# melihat tipe data
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [24]:
# melihat informasi df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [25]:
# melihat nilai df dalam array
df.values

array([[1, 0, 3, ..., 7.25, nan, 'S'],
       [2, 1, 1, ..., 71.2833, 'C85', 'C'],
       [3, 1, 3, ..., 7.925, nan, 'S'],
       ...,
       [889, 0, 3, ..., 23.45, nan, 'S'],
       [890, 1, 1, ..., 30.0, 'C148', 'C'],
       [891, 0, 3, ..., 7.75, nan, 'Q']], dtype=object)

In [26]:
# melihat dimensi
df.ndim

2

In [27]:
# melihat shape
df.shape

(891, 12)

In [28]:
# melihat size
df.size

10692

> Untuk melakukan indexing dan slicing, kita bisa terapkan cara yang sama persis dengan indexing atau slicing list pada python.

In [29]:
# contoh indexing kolom
df.Name.tail()

886                       Montvila, Rev. Juozas
887                Graham, Miss. Margaret Edith
888    Johnston, Miss. Catherine Helen "Carrie"
889                       Behr, Mr. Karl Howell
890                         Dooley, Mr. Patrick
Name: Name, dtype: object

In [30]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [31]:
# slicing >1 columns
df[["PassengerId", "Cabin", "Embarked", "Pclass"]].head()

PassengerId Cabin Embarked  Pclass
0            1   NaN        S       3
1            2   C85        C       1
2            3   NaN        S       3
3            4  C123        S       1
4            5   NaN        S       3

In [32]:
# indexing dengan loc (mulai:akhir)
df.loc[2:4]

PassengerId  Survived  Pclass  \
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
2                        Heikkinen, Miss. Laina  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket    Fare Cabin Embarked  
2  STON/O2. 3101282   7.925   NaN        S  
3            113803  53.100  C123        S  
4            373450   8.050   NaN        S

In [33]:
# indexing dengan iloc (mulai:akhir-1)
df.iloc[2:4]

PassengerId  Survived  Pclass  \
2            3         1       3   
3            4         1       1   

                                           Name     Sex   Age  SibSp  Parch  \
2                        Heikkinen, Miss. Laina  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   

             Ticket    Fare Cabin Embarked  
2  STON/O2. 3101282   7.925   NaN        S  
3            113803  53.100  C123        S

> Terdapat fungsi yang tersedia untuk melakukan manipulasi index dan value, yaitu `set_index()` (untuk membuat salah satu kolom menjad index), `sort_values()` (untuk melakukan sortir), dan `reset_index()` (untuk mereset index).

In [34]:
# contoh set nama menjadi index
df.set_index("Name")

PassengerId  Survived  \
Name                                                                        
Braund, Mr. Owen Harris                                       1         0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...            2         1   
Heikkinen, Miss. Laina                                        3         1   
Futrelle, Mrs. Jacques Heath (Lily May Peel)                  4         1   
Allen, Mr. William Henry                                      5         0   
...                                                         ...       ...   
Montvila, Rev. Juozas                                       887         0   
Graham, Miss. Margaret Edith                                888         1   
Johnston, Miss. Catherine Helen "Carrie"                    889         0   
Behr, Mr. Karl Howell                                       890         1   
Dooley, Mr. Patrick                                         891         0   

                                                    Pclass     Sex   Age  \
Name                                                                       
Braund, Mr. Owen Harris                                  3    male  22.0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...       1  female  38.0   
Heikkinen, Miss. Laina                                   3  female  26.0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)             1  female  35.0   
Allen, Mr. William Henry                                 3    male  35.0   
...                                                    ...     ...   ...   
Montvila, Rev. Juozas                                    2    male  27.0   
Graham, Miss. Margaret Edith                             1  female  19.0   
Johnston, Miss. Catherine Helen "Carrie"                 3  female   NaN   
Behr, Mr. Karl Howell                                    1    male  26.0   
Dooley, Mr. Patrick                                      3    male  32.0   

                                                    SibSp  Parch  \
Name                                                               
Braund, Mr. Owen Harris                                 1      0   
Cumings, Mrs. John Bradley (Florence Briggs Tha...      1      0   
Heikkinen, Miss. Laina                                  0      0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)            1      0   
Allen, Mr. William Henry                                0      0   
...                                                   ...    ...   
Montvila, Rev. Juozas                                   0      0   
Graham, Miss. Margaret Edith                            0      0   
Johnston, Miss. Catherine Helen "Carrie"                1      2   
Behr, Mr. Karl Howell                                   0      0   
Dooley, Mr. Patrick                                     0      0   

                                                              Ticket     Fare  \
Name                                                                            
Braund, Mr. Owen Harris                                    A/5 21171   7.2500   
Cumings, Mrs. John Bradley (Florence Briggs Tha...          PC 17599  71.2833   
Heikkinen, Miss. Laina                              STON/O2. 3101282   7.9250   
Futrelle, Mrs. Jacques Heath (Lily May Peel)                  113803  53.1000   
Allen, Mr. William Henry                                      373450   8.0500   
...                                                              ...      ...   
Montvila, Rev. Juozas                                         211536  13.0000   
Graham, Miss. Margaret Edith                                  112053  30.0000   
Johnston, Miss. Catherine Helen "Carrie"                  W./C. 6607  23.4500   
Behr, Mr. Karl Howell                                         111369  30.0000   
Dooley, Mr. Patrick                                           370376   7.7500   

                                                   Cabin Embarked  
Name                                                  

In [35]:
# contoh soft_values, ascending=False untuk descending, ascending=True untuk ascending
df.sort_values(by="Name", ascending=False)

PassengerId  Survived  Pclass                                   Name  \
868          869         0       3            van Melkebeke, Mr. Philemon   
153          154         0       3        van Billiard, Mr. Austin Blyler   
361          362         0       2              del Carlo, Mr. Sebastiano   
282          283         0       3              de Pelsmaeker, Mr. Alfons   
286          287         1       3                de Mulder, Mr. Theodore   
..           ...       ...     ...                                    ...   
874          875         1       2  Abelson, Mrs. Samuel (Hannah Wizosky)   
308          309         0       2                    Abelson, Mr. Samuel   
279          280         1       3       Abbott, Mrs. Stanton (Rosa Hunt)   
746          747         0       3            Abbott, Mr. Rossmore Edward   
845          846         0       3                    Abbing, Mr. Anthony   

        Sex   Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
868    male   NaN      0      0         345777   9.5000   NaN        S  
153    male  40.5      0      2       A/5. 851  14.5000   NaN        S  
361    male  29.0      1      0  SC/PARIS 2167  27.7208   NaN        C  
282    male  16.0      0      0         345778   9.5000   NaN        S  
286    male  30.0      0      0         345774   9.5000   NaN        S  
..      ...   ...    ...    ...            ...      ...   ...      ...  
874  female  28.0      1      0      P/PP 3381  24.0000   NaN        C  
308    male  30.0      1      0      P/PP 3381  24.0000   NaN        C  
279  female  35.0      1      1      C.A. 2673  20.2500   NaN        S  
746    male  16.0      1      1      C.A. 2673  20.2500   NaN        S  
845    male  42.0      0      0      C.A. 5547   7.5500   NaN        S  

[891 rows x 12 columns]

In [36]:
# contoh reset index
# drop=True untuk melakukanb drop pada kolom indexnya
df.set_index("Name").reset_index(drop=True)

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0              1         0       3    male  22.0      1      0   
1              2         1       1  female  38.0      1      0   
2              3         1       3  female  26.0      0      0   
3              4         1       1  female  35.0      1      0   
4              5         0       3    male  35.0      0      0   
..           ...       ...     ...     ...   ...    ...    ...   
886          887         0       2    male  27.0      0      0   
887          888         1       1  female  19.0      0      0   
888          889         0       3  female   NaN      1      2   
889          890         1       1    male  26.0      0      0   
890          891         0       3    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  
0           A/5 21171   7.2500   NaN        S  
1            PC 17599  71.2833   C85        C  
2    STON/O2. 3101282   7.9250   NaN        S  
3              113803  53.1000  C123        S  
4              373450   8.0500   NaN        S  
..                ...      ...   ...      ...  
886            211536  13.0000   NaN        S  
887            112053  30.0000   B42        S  
888        W./C. 6607  23.4500   NaN        S  
889            111369  30.0000  C148        C  
890            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

In [37]:
# cek hasil
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

### 4. Descriptive Statistic

> Dalam penggunaanya, kita dapat mengetahui hasil statistik deskriptif dengan menggunakan `describe()`. Walaupun yang terlihat disini hanya yang dalam bentuk numbers (int, float). Sehingga kadang kita perlu melakukan konversi datatype dan data cleansing terlebih dahulu agar hasil lebih tepat.

In [38]:
# contoh discribe
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

![](https://builtin.com/sites/www.builtin.com/files/styles/ckeditor_optimize/public/inline-images/national/data%2520types%2520examples.png)

1. Categorical = String
- Ordinal = data yang memiliki tingkat urutan (sd, smp, sma, kuliah)
- Nominal = data yang tidak memiliki tingat urutan (laki-laki, perempuan)

2. Numerical = Number
- Interval = data berbentuk jarak, tapi tidak ada nilai 0 (tinggi badan 100 - 200 cm)
- Ratio = data berbentuk jarak yang memiliki nilai 0 (kecepatan 0 - 100 km/jam)

> References: https://www150.statcan.gc.ca/n1/edu/power-pouvoir/ch8/5214817-eng.htm


> Dalam menafsirkan insight, kita perlu mengetahui data description terlebih dahulu dan sesuaikan dengan tipe data yang dimaksud agar insight yang kita beri sesuai dengan keadaan data pada situasi sebenarnya.

#### **Informasi data**:

- PassengerId: Id of every passenger.

- Survived: Indication whether passenger survived. 0 for yes and 1 for no.

- Pclass: One out of the 3 ticket classes: Class 1, Class 2 and Class 3. 

- Name: Name of passenger.

- Sex: Gender of passenger.

- Age: Age of passenger in years.

- SibSp: Number of siblings or spouses aboard.

- Parch: Number of parents or children aboard.

- Ticket: Ticket number of passenger.

- Fare: Indicating the fare.

- Cabin: Cabin number of passenger.

- Embarked: Port of embarkation.

In [39]:
# contoh discribe
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

**Contoh menafsirkan data**:


Dari data diatas, didapatkan nilai mean (rata-rata) = 29.6 atau dibulatkan menjadi 30, sehingga dapat dikatakan bahwa nilai rata-rata / sebagian besar umur passenger di data titanic ini adalah 30 tahun (asumsi distribusi gausian/normal).

Pada data diatas, juga didapatkan nilai standar deviasi (std / sebaran data) diangka 14,5 atau dibulatkan menjadi 15, sehingga dapat dikatakan, jika mengikuti asumsi gausian (normal), maka sebaran data berada pada mean-std dan mean+std. Atau 30-15 dan 30+15. Sehingga dengan kata lain sebaran umur pada data diatas sebagian besar adalah antara umur 15 - 45 tahun.

Sementara umur termuda adalah 0 tahun dan tetua adalah 80 tahun (nilai min dan max).

### 5. Filtering

> Pandas mempermudah melakukan modifikasi data. Kita dapat melakukan filtering sesuai kondisi yang kita inginkan dengan cara yang cukup mudah, yaitu dengan melakukan slicing pada kondisi yang diinginkan.

In [40]:
# contoh filter kondisi survive = 1
filt = df[(df["Survived"] == 1)]  # masukkan nilai
filt

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
875      0              2667   7.2250   NaN        C  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  

[342 rows x 12 columns]

In [41]:
# contoh lain filtering
df_female = df[df["Sex"] == "female"]
df_female

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
880          881         1       2   
882          883         0       3   
885          886         0       3   
887          888         1       1   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
880      1            230433  26.0000   NaN        S  
882      0              7552  10.5167   NaN        S  
885      5            382652  29.1250   NaN        Q  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  

[314 rows x 12 columns]

> Kita juga dapat melakukan filtering dengan lebih dari 1 kondisi dengan cara menambahkan tanda kurung pada tiap kondisi dan operator filternya `&` (dan) serta `|` (atau).

In [42]:
# contoh lebih dari 1 filter
df[(df["Sex"] == "female") & (df["Survived"] == 0) & (df["Age"] < 30)]  # operator and

PassengerId  Survived  Pclass  \
14            15         0       3   
24            25         0       3   
38            39         0       3   
41            42         0       2   
49            50         0       3   
71            72         0       3   
100          101         0       3   
111          112         0       3   
113          114         0       3   
114          115         0       3   
119          120         0       3   
147          148         0       3   
199          200         0       2   
205          206         0       3   
246          247         0       3   
251          252         0       3   
293          294         0       3   
297          298         0       1   
312          313         0       2   
374          375         0       3   
402          403         0       3   
404          405         0       3   
419          420         0       3   
423          424         0       3   
436          437         0       3   
474          475         0       3   
498          499         0       1   
501          502         0       3   
541          542         0       3   
542          543         0       3   
567          568         0       3   
617          618         0       3   
634          635         0       3   
642          643         0       3   
654          655         0       3   
702          703         0       3   
729          730         0       3   
807          808         0       3   
813          814         0       3   
816          817         0       3   
852          853         0       3   
882          883         0       3   

                                                  Name     Sex   Age  SibSp  \
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
24                       Palsson, Miss. Torborg Danira  female   8.0      3   
38                  Vander Planke, Miss. Augusta Maria  female  18.0      2   
41   Turpin, Mrs. William John Robert (Dorothy Ann ...  female  27.0      1   
49       Arnold-Franchi, Mrs. Josef (Josefine Franchi)  female  18.0      1   
71                          Goodwin, Miss. Lillian Amy  female  16.0      5   
100                            Petranec, Miss. Matilda  female  28.0      0   
111                               Zabour, Miss. Hileni  female  14.5      1   
113                            Jussila, Miss. Katriina  female  20.0      1   
114                              Attalah, Miss. Malake  female  17.0      0   
119                  Andersson, Miss. Ellis Anna Maria  female   2.0      4   
147                   Ford, Miss. Robina Maggie "Ruby"  female   9.0      2   
199             Yrois, Miss. Henriette ("Mrs Harbeck")  female  24.0      0   
205                         Strom, Miss. Telma Matilda  female   2.0      0   
246              Lindahl, Miss. Agda Thorilda Viktoria  female  25.0      0   
251         Strom, Mrs. Wilhelm (Elna Matilda Persson)  female  29.0      1   
293                                Haas, Miss. Aloisia  female  24.0      0   
297                       Allison, Miss. Helen Loraine  female   2.0      1   
312              Lahtinen, Mrs. William (Anna Sylfven)  female  26.0      1   
374                         Palsson, Miss. Stina Viola  female   3.0      3   
402                           Jussila, Miss. Mari Aina  female  21.0      1   
404                            Oreskovic, Miss. Marija  female  20.0      0   
419                          Van Impe, Miss. Catharina  female  10.0      0   
423  Danbom, Mrs. Ernst Gilbert (Anna Sigrid Maria ...  female  28.0      1   
436               Ford, Miss. Doolina Margaret "Daisy"  female  21.0      2   
474                        Strandberg, Miss. Ida Sofia  female  22.0      0   
498    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0      1   
501                                Canavan, Miss. Mary  female  21.0      0   
541               Andersson, Miss. Ingeborg Constanzia  female   9.0      4   


In [43]:
# contoh lebih dari 1 filter or
df[(df["Sex"] == "female") | (df["Survived"] == 0)]  # operator or

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
890      0            370376   7.7500   NaN        Q  

[782 rows x 12 columns]

In [44]:
# filter dengan where
cond = (df["Sex"] == "female") | (df["Survived"] == 0)
df.where(cond)

PassengerId  Survived  Pclass  \
0            1.0       0.0     3.0   
1            2.0       1.0     1.0   
2            3.0       1.0     3.0   
3            4.0       1.0     1.0   
4            5.0       0.0     3.0   
..           ...       ...     ...   
886        887.0       0.0     2.0   
887        888.0       1.0     1.0   
888        889.0       0.0     3.0   
889          NaN       NaN     NaN   
890        891.0       0.0     3.0   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0    1.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                               Heikkinen, Miss. Laina  female  26.0    0.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                             Allen, Mr. William Henry    male  35.0    0.0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0    0.0   
887                       Graham, Miss. Margaret Edith  female  19.0    0.0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN    1.0   
889                                                NaN     NaN   NaN    NaN   
890                                Dooley, Mr. Patrick    male  32.0    0.0   

     Parch            Ticket     Fare Cabin Embarked  
0      0.0         A/5 21171   7.2500   NaN        S  
1      0.0          PC 17599  71.2833   C85        C  
2      0.0  STON/O2. 3101282   7.9250   NaN        S  
3      0.0            113803  53.1000  C123        S  
4      0.0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886    0.0            211536  13.0000   NaN        S  
887    0.0            112053  30.0000   B42        S  
888    2.0        W./C. 6607  23.4500   NaN        S  
889    NaN               NaN      NaN   NaN      NaN  
890    0.0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### 6. Aggregation dan Combining

> Pandas mendukung fungsi agregasi dengan `groupby()`, setelah itu kita bisa tampilkan dengan fungsi agregasi seperti `sum()`, `mean()`, dst. Kita bisa melihat valuesnya dengan `value_counts()`.

In [45]:
# contoh groupby dengan value counts
df.groupby("Survived")["Sex"].value_counts()

Survived  Sex   
0         male      468
          female     81
1         female    233
          male      109
Name: Sex, dtype: int64

In [46]:
# contoh groupby dengan sum
df.groupby("Pclass")["Fare"].sum()

Pclass
1    18177.4125
2     3801.8417
3     6714.6951
Name: Fare, dtype: float64

In [47]:
# contoh groupby dengan max
df.groupby("Pclass")["Fare"].max()

Pclass
1    512.3292
2     73.5000
3     69.5500
Name: Fare, dtype: float64

In [48]:
# contoh groupby dengan mean dan di pass ke dataframe
df.groupby("Pclass", as_index=False)["Fare"].mean()

Pclass       Fare
0       1  84.154687
1       2  20.662183
2       3  13.675550

In [49]:
# contoh groupby >1 kolom
df.groupby(["Pclass", "Sex"], as_index=False)["Fare"].mean()

Pclass     Sex        Fare
0       1  female  106.125798
1       1    male   67.226127
2       2  female   21.970121
3       2    male   19.741782
4       3  female   16.118810
5       3    male   12.661633

> Selain itu, kita juga dapat menggabungkan data dengan `join()`, `append()`, `merge()`, `concat()`, `pivot_table()`, dan `cut()`. Konsepnya sama dengan native Python, SQL, Excel, dll.

> dokumentasi: https://pandas.pydata.org/docs/reference/general_functions.html


In [50]:
# contoh unique
df["Sex"].unique()

array(['male', 'female'], dtype=object)

In [51]:
# contoh nunique
df["Sex"].nunique()

2

In [52]:
# contoh join
# buat dataframe baru
pclass_mean = df.groupby("Pclass", as_index=False)["Fare"].mean()
pclass_sum = df.groupby("Pclass", as_index=False)["Fare"].sum()

In [53]:
pclass_mean.set_index("Pclass")

Fare
Pclass           
1       84.154687
2       20.662183
3       13.675550

In [54]:
pclass_sum.set_index("Pclass")

Fare
Pclass            
1       18177.4125
2        3801.8417
3        6714.6951

In [55]:
# contoh dengan join
pclass_mean.join(pclass_sum, on="Pclass", how="left",  # inner, right, left, outer
                lsuffix='_mean', rsuffix='_sum')  # menambahkan surffix pada nama kolom

Pclass_mean  Fare_mean  Pclass_sum   Fare_sum
0            1  84.154687         2.0  3801.8417
1            2  20.662183         3.0  6714.6951
2            3  13.675550         NaN        NaN

In [56]:
# contoh dengan join set index
pclass_mean.set_index("Pclass").join(pclass_sum.set_index("Pclass"),
                                     on="Pclass",
                                     how="left",  # inner, right, left, outer
                                     lsuffix='_mean',
                                     rsuffix='_sum').reset_index()  # menambahkan surffix pada nama kolom

Pclass  Fare_mean    Fare_sum
0       1  84.154687  18177.4125
1       2  20.662183   3801.8417
2       3  13.675550   6714.6951

In [57]:
# contoh merge, alternatif join (bisa cross join di versi pandas 1.2.0 keatas)
pd.merge(pclass_mean, pclass_sum, on="Pclass", how="left")

Pclass     Fare_x      Fare_y
0       1  84.154687  18177.4125
1       2  20.662183   3801.8417
2       3  13.675550   6714.6951

In [58]:
# contoh dengan append, sifatnya sama seperti list
pclass_mean.append(pclass_sum)

C:\Users\Asus\AppData\Local\Temp\ipykernel_16672\4046605237.py:2: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  pclass_mean.append(pclass_sum)


Pclass          Fare
0       1     84.154687
1       2     20.662183
2       3     13.675550
0       1  18177.412500
1       2   3801.841700
2       3   6714.695100

In [59]:
# contoh dengan concat, mirip seperti append jika axis = 0
print(pd.concat([pclass_mean, pclass_sum], axis=0))
print("\n")  # axis = 1 untuk horizontal, 0 untuk vertikal
pd.concat([pclass_mean, pclass_sum], axis=1)

   Pclass          Fare
0       1     84.154687
1       2     20.662183
2       3     13.675550
0       1  18177.412500
1       2   3801.841700
2       3   6714.695100




Pclass       Fare  Pclass        Fare
0       1  84.154687       1  18177.4125
1       2  20.662183       2   3801.8417
2       3  13.675550       3   6714.6951

In [60]:
# contoh pivot table
df.pivot_table("Fare", index=["Pclass"])  # bisa .reset_index()

Fare
Pclass           
1       84.154687
2       20.662183
3       13.675550

In [61]:
# contoh pivot table dengan index > 1
df.pivot_table("Fare",  # kolom utama (ambil datanya)
               index=["Sex", "Survived", "Embarked"],  # index yang diinginkan
               columns=["Pclass"],  # kolom paling atas untuk pivot
               aggfunc={"Fare": np.mean})  # aggfunc bisa np.mean, np.sum, dll

Pclass                             1          2          3
Sex    Survived Embarked                                  
female 0        C          28.712500        NaN  14.653125
                Q                NaN        NaN  10.904633
                S         151.550000  18.250000  21.969018
       1        C         117.710019  25.268457  14.717220
                Q          90.000000  12.350000  10.084033
                S          96.743298  22.272951  13.171842
male   0        C          80.847160  25.418225   8.720842
                Q          90.000000  12.350000  11.841550
                S          53.563316  18.997561  12.758689
       1        C         112.197806  25.433350  11.435840
                Q                NaN        NaN  12.916667
                S          51.832739  20.516667  17.033450

> Dokumentasi: https://www.pythonprogramming.in/numpy-aggregate-and-statistical-functions.html

In [62]:
# contoh cut, untuk membagi bins (membagi ke kelompoknya, contoh dari 20-40, 40-60)
pd.cut(np.array([1, 7, 5, 4, 6, 4]), 3)

[(0.994, 3.0], (5.0, 7.0], (3.0, 5.0], (3.0, 5.0], (5.0, 7.0], (3.0, 5.0]]
Categories (3, interval[float64, right]): [(0.994, 3.0] < (3.0, 5.0] < (5.0, 7.0]]

In [63]:
# contoh melakukan cut untuk membagi interval umur
age_cut = pd.cut(df["Age"], [0, 20, 40, 60, 80])
age_cut

0      (20.0, 40.0]
1      (20.0, 40.0]
2      (20.0, 40.0]
3      (20.0, 40.0]
4      (20.0, 40.0]
           ...     
886    (20.0, 40.0]
887     (0.0, 20.0]
888             NaN
889    (20.0, 40.0]
890    (20.0, 40.0]
Name: Age, Length: 891, dtype: category
Categories (4, interval[int64, right]): [(0, 20] < (20, 40] < (40, 60] < (60, 80]]

In [64]:
# cek interval
age_cut.unique()

[(20.0, 40.0], NaN, (40.0, 60.0], (0.0, 20.0], (60.0, 80.0]]
Categories (4, interval[int64, right]): [(0, 20] < (20, 40] < (40, 60] < (60, 80]]

In [65]:
# cek iloc tertentu
print(df.loc[700])
print("\nbins:", age_cut[700])

PassengerId                                                  701
Survived                                                       1
Pclass                                                         1
Name           Astor, Mrs. John Jacob (Madeleine Talmadge Force)
Sex                                                       female
Age                                                         18.0
SibSp                                                          1
Parch                                                          0
Ticket                                                  PC 17757
Fare                                                     227.525
Cabin                                                    C62 C64
Embarked                                                       C
Name: 700, dtype: object

bins: (0, 20]


In [66]:
# menggabungkan cut dan pivot table
df.pivot_table("Survived",  # kolom utama (ambil datanya)
               index=["Sex", age_cut],  # index yang diinginkan
               columns=["Pclass"],  # kolom paling atas untuk pivot
               aggfunc={"Survived": np.mean})  # aggfunc bisa np.mean, np.sum, dll

Pclass                  1         2         3
Sex    Age                                   
female (0, 20]   0.928571  1.000000  0.510638
       (20, 40]  0.977778  0.911111  0.478261
       (40, 60]  0.958333  0.846154  0.000000
       (60, 80]  1.000000       NaN  1.000000
male   (0, 20]   0.571429  0.526316  0.197368
       (20, 40]  0.500000  0.050847  0.143836
       (40, 60]  0.342105  0.055556  0.074074
       (60, 80]  0.083333  0.333333  0.000000

Contoh menafsirkan data diatas untuk penumpang wanita dengan umur dibawah 20 tahun, dan class 1 yang selamat:

---

1. Bisa dikatakan dari data diatas, bahwa *survivalibily* dari gender *female* pada kategori umur 0 - 20 tahun terhitung sangat rendah untuk *class* 1 dan 2, sedangkan untuk penumpang *class* 3 memiliki *survivability* sebesar +-50%.

2. Melalui data diatas, jika kita ambil kesimpulan secara garis besar, bahwa apabila rate kematian > 80%, maka *survivability*-nya termasuk rendah. Maka berdasarkan data, kebanyakan *female* yang berada pada Titanic tersebut memiliki *survivability* yang rendah, kecuali mereka yang berada pada *class* 3, diusia < 60 tahun.

3. Sementara pada *male*, tingkat kematiannya < 50% pada hampir semua *passanger class* dari segala kategori umur, kecuali mereka yang berada di *passenger class* 1 dan 2, dengan kategori umur 0 - 20 tahun dan *passenger class* 1, dengan kategori umur 20 - 40 tahun.

In [67]:
df[(df["Sex"] == "female") & (df["Age"] < 20) & (df["Pclass"] == 1)]["Survived"].value_counts()

1    13
0     1
Name: Survived, dtype: int64

> persentase yang tidak selamat (wanita, umur dibawah 20 thn, dan pclass = 1) 13/14 = 0.92%

### 7. Data Type Conversion and Missing Value Handling

> Biasanya tahap awal sebelum melakukan data analysis adalah data cleansing terlebih dahulu. Dalam proses ini, kita bisa lakukan inspeksi terhadap masing-masing tipe data pada kolom di dataframe, jika ada yang belum tepat (sesuai data description) maka kita harus melakukan handling yaitu melakukan konversi ke datatype yang seharusnya sesuai tipe data python (str, int, float, dIl). Pada Pandas hal ini bisa dengan mudah kita lakukan menggunakan `astype()`.

In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


> misalkan pasengerId tidak dapat dihitung, survived bukan dalam bentuk int, dll

In [69]:
# cek missing values pada tiap kolom
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

> Contoh, setelah melihat data type dan data dict, kita ketahui Pclass seharusnya bukan int, kita harus merubah ini ke str dengan `astype()`.

In [70]:
# contoh data type conversion
df["PassengerId"] = df["PassengerId"].astype(str)

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    object 
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 83.7+ KB


In [72]:
df["Survived"].replace(0, "Survive")

0      Survive
1            1
2            1
3            1
4      Survive
        ...   
886    Survive
887          1
888    Survive
889          1
890    Survive
Name: Survived, Length: 891, dtype: object

In [73]:
df["Survived"].replace({0: "Survive", 1: "Death"})

0      Survive
1        Death
2        Death
3        Death
4      Survive
        ...   
886    Survive
887      Death
888    Survive
889      Death
890    Survive
Name: Survived, Length: 891, dtype: object

> Kadang saat melakukan konversi, kita menemukan exception, yang harus dilakukan adalah kita harus melihat exception-nya dan solve.

In [74]:
# contoh melakukan konversi terhadap kolom Age ke int64 (akan error)
# df["Age"] = df["Age"].astyoe("int64")

In [75]:
# kita cek dulu penyebab errornya
for i, j in enumerate(df["Age"]):
    try:
        j = int(j)  # coba rubah manual ke int
    except:
        print(i, j)  # print index dan valuenya

5 nan
17 nan
19 nan
26 nan
28 nan
29 nan
31 nan
32 nan
36 nan
42 nan
45 nan
46 nan
47 nan
48 nan
55 nan
64 nan
65 nan
76 nan
77 nan
82 nan
87 nan
95 nan
101 nan
107 nan
109 nan
121 nan
126 nan
128 nan
140 nan
154 nan
158 nan
159 nan
166 nan
168 nan
176 nan
180 nan
181 nan
185 nan
186 nan
196 nan
198 nan
201 nan
214 nan
223 nan
229 nan
235 nan
240 nan
241 nan
250 nan
256 nan
260 nan
264 nan
270 nan
274 nan
277 nan
284 nan
295 nan
298 nan
300 nan
301 nan
303 nan
304 nan
306 nan
324 nan
330 nan
334 nan
335 nan
347 nan
351 nan
354 nan
358 nan
359 nan
364 nan
367 nan
368 nan
375 nan
384 nan
388 nan
409 nan
410 nan
411 nan
413 nan
415 nan
420 nan
425 nan
428 nan
431 nan
444 nan
451 nan
454 nan
457 nan
459 nan
464 nan
466 nan
468 nan
470 nan
475 nan
481 nan
485 nan
490 nan
495 nan
497 nan
502 nan
507 nan
511 nan
517 nan
522 nan
524 nan
527 nan
531 nan
533 nan
538 nan
547 nan
552 nan
557 nan
560 nan
563 nan
564 nan
568 nan
573 nan
578 nan
584 nan
589 nan
593 nan
596 nan
598 nan
601 nan
602 nan

In [76]:
# cek salah satu
df["Age"][206]

32.0

In [77]:
df.isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

#### **Terdapat 3 jenis missing value:**
---
1. **MCAR (Missing Completely at Random)** atau data yang hilang tetapi tidak terkait dengan suatu variable, sehingga untuk kebanyakan kasus hal ini terjadi karena error, bisa human error atau corrupt pada data. **Cara handling nya adalah dengan drop, walaupun tetap disarankan diisi**. (hilangnya < 3%, sehingga tidak masalah untuk di drop)

2. **MAR (Missing at Random)** data yang hilang kemungkinan besar terkait dengan variablenya, atau biasanya dalam proses sampling / pengumpulan data, responden sengaja tidak menjawab pertanyaan untuk variable ini, entah mungkin karena memang privasi atau hat lain, sehingga nilainya menjadi kosong. **Cara handling nya adalah dengan missing values imputation**. (hilangnya > 10%, sehingga tidak boleh di drop karena akan berpengaruh, biasanya memakai mean (ada proxy, contohnya ada data gaji pekerja di bidang IT dan lama kerja 5 tahun yang hilang, maka kita harus memakai rata-rata dari data pekerja IT yang lama kerjanya 5 tahun juga))

3. **MNAR (Missing Not at Random)** merupakan data yang hilang memang bukan karena error atau terkait dengan variable tertentu seperti pada MCAR atau MAR. Terjadinya missing data pada suatu variabel berkaitan dengan variabel itu sendiri, sehingga ini tidak bisa diprediksi dari variabel lain pada suatu dataset. Data yang hilang pada hal ini memang bisa jadi adalah nilai dari suatu data (value = Null). **Cara handling nya adalah dengan mengubah ke dalam isi yang dimaksud, dalam case tertentu Null ini dapat diisi dengan 0 atau unknown**. (isi data tersebut memang null, bukan tidak terisi)

In [78]:
# copy dataframe terlebih dahulu
df4 = df.copy()

#### Untuk handling missing values, terutama pada MCAR dan MAR dapat dilakukan dengan beberapa cara:

1. Dapat langsung drop data yang hilang pada kolom  (definisikan kolomnya)

2. Dapat buang kolom yang memiliki banyak null values
 
3. Dapat kita isi sebagaimana maksudnya

> **DROP NULL VALUES** </br>
Drop null values memiliki 2 cara pendekatan, yaitu gunakan filter + `notna()` atau gunakan `dropna()` yang keduanya sama-sama memiliki fungsi yang sama, namun karakteristik yang berbeda.

In [79]:
# filter + notna
df5 = df4[df4["Age"].notna()].reset_index(drop=True)  # gunakan notna dan reset index

In [80]:
# cek df5
df5.shape

(714, 12)

In [81]:
df5.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          529
Embarked         2
dtype: int64

In [82]:
# drop missing value pada dataframe
df6 = df4.dropna()

In [83]:
# cek df6
df6.shape

(183, 12)

In [84]:
df6.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [85]:
# lihat shape data setelah drop
print("Original shape:", df.shape)
print("notna pada kolom 'Age':", df5.shape)
print("dropna pada kolom 'Age':", df6.shape)

Original shape: (891, 12)
notna pada kolom 'Age': (714, 12)
dropna pada kolom 'Age': (183, 12)


Dokumentasi:
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html

- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.notna.html


In [86]:
# drop kolom dengan fungsi drop
df6 = df4.drop(columns="Age")

In [87]:
df6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    object 
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    float64
 9   Cabin        204 non-null    object 
 10  Embarked     889 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 76.7+ KB


**FILL / REPLACE NULL VALUES** <br>

---
Kita dapat melakukan fill pada Null values dengan fungsi `fillna()`, `ffill()`, `bfill()` dan `interpolate()`. Selain itu kita dapat juga gunakan replace() sebagai
alternatif lain atas fungsi replace ini.
Lebih lanjut: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html

In [88]:
# copy dataframe
df7 = df.copy()

In [89]:
# contoh dengan fillna
df7["Age"] = df7["Age"].fillna(0)  # mengganti seluruh nilai NUll dengan 0

In [90]:
# cek hasil fill
df7.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [91]:
# bandingkan hasil fillna
pd.DataFrame({"before": df["Age"].head(10), "after": df7["Age"].head(10)})

before  after
0    22.0   22.0
1    38.0   38.0
2    26.0   26.0
3    35.0   35.0
4    35.0   35.0
5     NaN    0.0
6    54.0   54.0
7     2.0    2.0
8    27.0   27.0
9    14.0   14.0

In [92]:
df7 = df.copy()

In [93]:
# contoh dengan ffill
df7["Age"] = df7["Age"].ffill()

In [94]:
# cek nill values
df7.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [95]:
# bandingkan dengan ffill
pd.DataFrame({"before": df["Age"].head(10), "after": df7["Age"].head(10)})

before  after
0    22.0   22.0
1    38.0   38.0
2    26.0   26.0
3    35.0   35.0
4    35.0   35.0
5     NaN   35.0
6    54.0   54.0
7     2.0    2.0
8    27.0   27.0
9    14.0   14.0

In [96]:
df7 = df.copy()

In [97]:
# contoh dengan bfill
df7["Age"] = df7["Age"].bfill()

In [98]:
# cek nill values
df7.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [99]:
# bandingkan dengan dfill
pd.DataFrame({"before": df["Age"].head(10), "after": df7["Age"].head(10)})

before  after
0    22.0   22.0
1    38.0   38.0
2    26.0   26.0
3    35.0   35.0
4    35.0   35.0
5     NaN   54.0
6    54.0   54.0
7     2.0    2.0
8    27.0   27.0
9    14.0   14.0

In [100]:
# copy dataframe
df7 = df.copy()

In [101]:
# contoh dengan fungsi replace
# gunakan numpy untuk mengidentifikasi null value
df7["Age"] = df7["Age"].replace(np.nan, "salah")

In [105]:
# bandingkan dengan dfill
pd.DataFrame({"before": df["Age"].head(10), "after": df7["Age"].head(10)})

before  after
0    22.0   22.0
1    38.0   38.0
2    26.0   26.0
3    35.0   35.0
4    35.0   35.0
5     NaN  salah
6    54.0   54.0
7     2.0    2.0
8    27.0   27.0
9    14.0   14.0

In [104]:
df7.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

> Dalam kebanyakan kasus, kita perlu melakukan data understanding dan analysis terlebih dahulu sebelum akhirnya kita mengisi Null values dengan metode yang dicontohkan diatas.

---
Jika kembali pada teori missing values, maka kita harus identifikasi lagi Null values ini adalah termasuk MCAR, MAR, atau MNAR.

- Jika Null termasuk MNAR, maka kita harus mengisi Null (dengan metode fill apapun) dengan nilai O, 'unknown'; dst yang merupakan representasi dari kosongnya data, tanpa harus merubah makna dari kekosongannya tersebut.

- Jika Null termasuk MCAR, maka kita bisa drop saja Null nya, gunakan  `ffi11()`, maupun `bfil1()` karena pada Null values ini masih aman. Biasanya data yang hilang jumlahnya kecil <3%.

- JIka Null termasuk MAR, maka kita harus isi Null tersebut dengan se-presisi mungkin, sehingga maksud dari data tersebut akan tetap tersampaikan dengan baik. Misal dengan melakukan filter data sesuai dengan informasi yang tersirat pada kolom lain dari dataset yang sama, lalu diisi Null values tersebut dengan mean, median atau modenya. Metode ini sering disebut Proxying.

In [106]:
# copy dataframe
df9 = df.copy()

In [108]:
# mencari korelasi
df9.corr()

Survived    Pclass       Age     SibSp     Parch      Fare
Survived  1.000000 -0.338481 -0.077221 -0.035322  0.081629  0.257307
Pclass   -0.338481  1.000000 -0.369226  0.083081  0.018443 -0.549500
Age      -0.077221 -0.369226  1.000000 -0.308247 -0.189119  0.096067
SibSp    -0.035322  0.083081 -0.308247  1.000000  0.414838  0.159651
Parch     0.081629  0.018443 -0.189119  0.414838  1.000000  0.216225
Fare      0.257307 -0.549500  0.096067  0.159651  0.216225  1.000000

In [109]:
# contoh lakukan filter null values dan pecah ke dalam df (variabel) baru
df_null = df9[df9["Age"].isna()]
df_not_null = df9[df9["Age"].notna()]

In [110]:
# cek data null
df_null.head()

PassengerId  Survived  Pclass                           Name     Sex  Age  \
5            6         0       3               Moran, Mr. James    male  NaN   
17          18         1       2   Williams, Mr. Charles Eugene    male  NaN   
19          20         1       3        Masselmani, Mrs. Fatima  female  NaN   
26          27         0       3        Emir, Mr. Farred Chehab    male  NaN   
28          29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  NaN   

    SibSp  Parch  Ticket     Fare Cabin Embarked  
5       0      0  330877   8.4583   NaN        Q  
17      0      0  244373  13.0000   NaN        S  
19      0      0    2649   7.2250   NaN        C  
26      0      0    2631   7.2250   NaN        C  
28      0      0  330959   7.8792   NaN        Q

In [111]:
# cek data notnull
df_not_null.head()

PassengerId  Survived  Pclass  \
0           1         0       3   
1           2         1       1   
2           3         1       3   
3           4         1       1   
4           5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

> Misal dari data diatas kita putuskan secara data understanding, kita akan melakukan proxy ke kolom Sex dan Pclass untuk mencari nilai kolom Age. Dan dimisalkan, nilai yang akan kita pilih untuk pengisian adalah mean.

In [114]:
# gunakan groupby untuk langsung mengetahui nilai mean
df_not_null.groupby(["Sex", "Pclass"], as_index=False)["Age"].mean()

Sex  Pclass        Age
0  female       1  34.611765
1  female       2  28.722973
2  female       3  21.750000
3    male       1  41.281386
4    male       2  30.740707
5    male       3  26.507589

Dari tabel di atas, kita dapat langsung mengetahui berapa nilai yang "pantas" untuk kita isi di kolom Age, ketika:
- Sex = female, dan Pclass = 1, maka nilai Age = 35th (34.6 di bulatkan keatas)

- Sex = female, dan Pclass = 2, maka nilai Age = 29th (28.7 di bulatkan keatas)

- dst

Ada beberapa cara untuk mengimplementasi hal ini, tetapi umumnya bisa dengan loop atau dengan filter manual.

In [115]:
# Cara 1, dengan loop
age = []
for sex, pclass in zip(df_null["Sex"], df_null["Pclass"]):
    if sex == "female":
        if pclass == 1:
            age.append(35)
        elif pclass == 2:
            age.append(29)
        else:
            age.append(22)
    else:
        if pclass == 1:
            age.append(41)
        elif pclass == 2:
            age.append(31)
        else:
            age.append(27)

In [117]:
import warnings
warnings.filterwarnings("ignore")

In [118]:
# masukkan hasil ke data (ignore error)
df_null["Age"] = age

In [122]:
# cek hasil pengisian
df_null[(df_null["Sex"] == "female") & (df_null["Pclass"] == 1)]

PassengerId  Survived  Pclass  \
31           32         1       1   
166         167         1       1   
256         257         1       1   
306         307         1       1   
334         335         1       1   
375         376         1       1   
457         458         1       1   
669         670         1       1   
849         850         1       1   

                                                  Name     Sex  Age  SibSp  \
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   35      1   
166             Chibnall, Mrs. (Edith Martha Bowerman)  female   35      0   
256                     Thorne, Mrs. Gertrude Maybelle  female   35      0   
306                            Fleming, Miss. Margaret  female   35      0   
334  Frauenthal, Mrs. Henry William (Clara Heinshei...  female   35      1   
375              Meyer, Mrs. Edgar Joseph (Leila Saks)  female   35      1   
457                  Kenyon, Mrs. Frederick R (Marion)  female   35      1   
669  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  female   35      1   
849       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female   35      1   

     Parch    Ticket      Fare Cabin Embarked  
31       0  PC 17569  146.5208   B78        C  
166      1    113505   55.0000   E33        S  
256      0  PC 17585   79.2000   NaN        C  
306      0     17421  110.8833   NaN        C  
334      0  PC 17611  133.6500   NaN        S  
375      0  PC 17604   82.1708   NaN        C  
457      0     17464   51.8625   D21        S  
669      0     19996   52.0000  C126        S  
849      0     17453   89.1042   C92        C

In [123]:
# join kembali data dengan data null, not null
df_merged = pd.concat([df_not_null, df_null])

In [124]:
df_merged.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   
10          11         1       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S

In [125]:
# cara 2
# copy dan susun dataframe contoh
df9 = df.copy()
df_null = df9[df9["Age"].isna()]
df_not_null = df9[df9["Age"].notna()]

In [127]:
# contoh dengan filter manual, masukkan dalam dataframe (variabel) baru
f1 = df_null[(df_null["Sex"] == "female") & (df_null["Pclass"] == 1)]
f2 = df_null[(df_null["Sex"] == "female") & (df_null["Pclass"] == 2)]
f3 = df_null[(df_null["Sex"] == "female") & (df_null["Pclass"] == 3)]
f4 = df_null[(df_null["Sex"] == "male") & (df_null["Pclass"] == 1)]
f5 = df_null[(df_null["Sex"] == "male") & (df_null["Pclass"] == 2)]
f6 = df_null[(df_null["Sex"] == "male") & (df_null["Pclass"] == 3)]

In [128]:
# apply filter manual
f1["Age"] = f1["Age"].fillna(35)
f2["Age"] = f2["Age"].fillna(29)
f3["Age"] = f3["Age"].fillna(22)
f4["Age"] = f4["Age"].fillna(41)
f5["Age"] = f5["Age"].fillna(31)
f6["Age"] = f6["Age"].fillna(27)

In [129]:
# join semua null data
df_null = pd.concat([f1, f2, f3, f4, f5, f6])

In [131]:
# cek hasil pengisian
df_null[(df_null["Sex"] == "female") & (df_null["Pclass"] == 1)].head(10)

PassengerId  Survived  Pclass  \
31           32         1       1   
166         167         1       1   
256         257         1       1   
306         307         1       1   
334         335         1       1   
375         376         1       1   
457         458         1       1   
669         670         1       1   
849         850         1       1   

                                                  Name     Sex   Age  SibSp  \
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female  35.0      1   
166             Chibnall, Mrs. (Edith Martha Bowerman)  female  35.0      0   
256                     Thorne, Mrs. Gertrude Maybelle  female  35.0      0   
306                            Fleming, Miss. Margaret  female  35.0      0   
334  Frauenthal, Mrs. Henry William (Clara Heinshei...  female  35.0      1   
375              Meyer, Mrs. Edgar Joseph (Leila Saks)  female  35.0      1   
457                  Kenyon, Mrs. Frederick R (Marion)  female  35.0      1   
669  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  female  35.0      1   
849       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female  35.0      1   

     Parch    Ticket      Fare Cabin Embarked  
31       0  PC 17569  146.5208   B78        C  
166      1    113505   55.0000   E33        S  
256      0  PC 17585   79.2000   NaN        C  
306      0     17421  110.8833   NaN        C  
334      0  PC 17611  133.6500   NaN        S  
375      0  PC 17604   82.1708   NaN        C  
457      0     17464   51.8625   D21        S  
669      0     19996   52.0000  C126        S  
849      0     17453   89.1042   C92        C

In [132]:
# join kembali data dengan data null, not null
df_merged = pd.concat([df_not_null, df_null])

In [133]:
df_merged.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   
10          11         1       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S

In [134]:
# rubah ke int
df_merged["Age"] = df_merged["Age"].astype("int64")
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 878
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    object 
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    int64  
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(1), int64(5), object(6)
memory usage: 90.5+ KB


In [135]:
df_merged.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  891.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.367003    0.523008    0.381594   32.204208
std      0.486592    0.836071   13.268658    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.000000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   22.000000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   27.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   36.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200